### PR TITLE
MODLOGIN-205. Call users API to get home tenant

### PR DESCRIPTION
**Pull request by SerhiiNosko**
## Purpose
[MODLOGIN-205](https://issues.folio.org/browse/MODLOGIN-205) - Add call to user_tenant table to get primary affiliation during login process
## Approach
- Updated `LoginCredentials` dto schema with a new field: `tenantId`
- New user service to get the home tenant by (`username` or `userId`) and `tenantId` from mod-users' `/user-tenants`. Also moved LoginAPI's `lookupUser()` to the new service.
- New unit tests for consortia logins
- Moved the unit tests for `lookupUser()` from `LoginAPITest` to `UserServiceImplTest`

Specs

- /user-tenants is called even if tenantId is not specified

- if there are multiple matching records, return an error with the list

- if /user-tenants returns an empty list, continue as before without an error at this point

### CODE REVIEW

File: src/main/java/org/folio/rest/impl/InitAPIs.java [hunk 1]
```diff
@@ -44,6 +46,9 @@ public void init(Vertx vertx, Context context, Handler<AsyncResult<Boolean>> res
       new ServiceBinder(vertx)
         .setAddress(EVENT_CONFIG_PROXY_CONFIG_ADDRESS)
         .register(ConfigurationService.class, ConfigurationService.create(vertx));
+      new ServiceBinder(vertx)
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/rest/impl/InitAPIs.java:49]: just for my information, why are these needed?

I was expecting to just make an api call to mod-users.

File: src/main/java/org/folio/rest/impl/LoginAPI.java [hunk 2]
```diff
@@ -288,184 +268,255 @@ public void postAuthnLogin(String userAgent, String xForwardedFor,
...
+            } else {
+              promise.complete(tenantId);
+            }
+          } else if (userTenants.size() > 1) {
+            List<String> matchingTenants = userTenants.stream()
+                .map(UserTenant::getTenantId)
+                .collect(toList());
+            logger.debug(MULTIPLE_MATCHING_USERS_LOG, credentials.getUsername(), credentials.getUserId(),
+                matchingTenants.toString());
+            asyncResultHandler.handle(Future.succeededFuture(PostAuthnLoginResponse.respond422WithApplicationJson(
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/rest/impl/LoginAPI.java:332]: It looks like a list of tenant Ids are returned as a response. Is the UI supposed to get tenant names from mod-consortia?

File: src/main/java/org/folio/rest/impl/LoginAPI.java [hunk 3]
```diff
@@ -288,184 +268,255 @@ public void postAuthnLogin(String userAgent, String xForwardedFor,
...
+      LoginConfigUtils.encodeJsonHeaders(okapiHeaders), ar -> {
+        if (ar.failed()) {
+          promise.fail(ar.cause());
+          return;
+        }
+        try {
+          List<UserTenant> userTenants = ar.result().stream()
+            .map(obj -> ((JsonObject)obj).mapTo(UserTenant.class))
+            .collect(toList());
+          if (userTenants.isEmpty()) {
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/rest/impl/LoginAPI.java:316]: Please add some comments in sections of if statement to summarize the logic. The three main sections are userTenants is empty, userTenants has more than one & the last case.

File: src/main/java/org/folio/rest/impl/LoginAPI.java
Comment (okolawole-ebsco) [inline src/main/java/org/folio/rest/impl/LoginAPI.java:1]: Good job for attempting to break up the very large login method!

File: src/main/java/org/folio/services/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+  private Future<JsonArray> getUserTenants(Map<String,String> headers, String currentTenantId, String username,
+      String userId, String requestedTenantId) {
+    Map<String,String> okapiHeaders = new CaseInsensitiveMap<>(headers);
+    String okapiUrl = okapiHeaders.get(XOkapiHeaders.URL);
+    String okapiToken = okapiHeaders.get(XOkapiHeaders.TOKEN);
+    String requestUrl = String.format(REQUEST_URL_TEMPLATE, okapiUrl, USER_TENANTS_URI_PATH);
+
+    HttpRequest<Buffer> request = WebClientFactory.getWebClient(vertx).getAbs(requestUrl);
+    if (userId != null)
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/services/impl/UserServiceImpl.java:120]: if userId and username are supplied, which would take precendence in mod-users?

File: src/main/java/org/folio/rest/impl/LoginAPI.java [hunk 5]
```diff
@@ -288,184 +268,255 @@ public void postAuthnLogin(String userAgent, String xForwardedFor,
...
-      userVerified.onComplete(verifyResult -> {
-        if (verifyResult.failed()) {
-          String errMsg = "Error verifying user existence: " + verifyResult
-              .cause().getLocalizedMessage();
-          logger.error(errMsg);
-          asyncResultHandler.handle(Future.succeededFuture(
-              PostAuthnLoginResponse.respond422WithApplicationJson(
-                  getErrors(errMsg, CODE_USERNAME_INCORRECT, new ImmutablePair<>(USERNAME, entity.getUsername())))
-          ));
+      handleCrossTenantLogin(entity, tenantId, okapiHeaders, asyncResultHandler)
```
Comment (okolawole-ebsco) [inline src/main/java/org/folio/rest/impl/LoginAPI.java:271]: Even regular logins will execute a method called `handleCrossTenantLogin`? Reads like it shouldn't be the case.

File: ramls/userTenant.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    },
+    "tenantId": {
+      "description": "Tenant identifier",
+      "type": "string"
+    },
+    "tenantName": {
+      "description": "Tenant name",
+      "type": "string"
+    },
+    "isPrimary": {
```
Comment (SerhiiNosko) [inline ramls/userTenant.json:26]: All records will be primary records in mod-users so we don't need this field

File: src/main/java/org/folio/rest/impl/InitAPIs.java (hunk 1 above)
Comment (damien-git) [reply src/main/java/org/folio/rest/impl/InitAPIs.java:49]: The user service is not necessary, but it was part of the AC and it doesn't hurt to encapsulate mod-users calls into a service. I am not sure about the specific kind of service used it mod-login, I just tried to keep the new one consistent with the others.

File: src/main/java/org/folio/rest/impl/LoginAPI.java (hunk 2 above)
Comment (damien-git) [reply src/main/java/org/folio/rest/impl/LoginAPI.java:332]: "Tenant ids" (or simply "tenants") are not UUIDs, they are names like "diku". I don't know if there are more readable names associated with them. If there are, then I guess the UI would need them.

File: src/main/java/org/folio/rest/impl/LoginAPI.java (hunk 3 above)
Comment (damien-git) [reply src/main/java/org/folio/rest/impl/LoginAPI.java:316]: I will add comments. The last case is for a single matching user-tenant association.

File: src/main/java/org/folio/services/impl/UserServiceImpl.java (hunk 4 above)
Comment (damien-git) [reply src/main/java/org/folio/services/impl/UserServiceImpl.java:120]: I both are supplied, both have to match for success. So, there is no precedence.

File: src/main/java/org/folio/rest/impl/LoginAPI.java (hunk 5 above)
Comment (damien-git) [reply src/main/java/org/folio/rest/impl/LoginAPI.java:271]: Yes, Serhii suggested always calling mod-users' /user-tenants and calling it cross-tenant rather than consortia. What would you prefer ?

File: ramls/userTenant.json (hunk 6 above)
Comment (damien-git) [reply ramls/userTenant.json:26]: OK

File: ramls/userTenant.json [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      "type": "string"
+    },
+    "tenantId": {
+      "description": "Tenant identifier",
+      "type": "string"
+    },
+    "tenantName": {
+      "description": "Tenant name",
+      "type": "string"
+    },
```
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply ramls/userTenant.json:25]: Thanks, we don't need tenant name in this table, we need only tenantId, so we will delete it. By the way, we have own API to manage tenants in consortia [[Link](https://s3.amazonaws.com/foliodocs/api/mod-consortia/s/tenants.htm)](https://s3.amazonaws.com/foliodocs/api/mod-consortia/s/tenants.html)
In our requirements we have possibility to edit tenant name and we are editing only tenant table in mod-consortia. In some future I think we need to call endpoint that @julianladisch provided from mod-consortia to edit tenant name in okapi to make things consistent
CC: @okolawole-ebsco

[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply ramls/userTenant.json:25]: Regarding technical perspective and best practices is it okay to call okapi endpoint from mod-consortia?

File: ramls/userTenant.json [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    "userId": {
+      "description": "User identifier",
+      "type": "string"
+    },
+    "username": {
+      "description": "Username",
+      "type": "string"
+    },
```
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply ramls/userTenant.json:17]: This table in mod-users will have always username and userId populated, we will add both fields to 'required' array

[the remark this answers is not in the document]
Comment (damien-git) [reply ramls/userTenant.json:17]: I think all the remaining fields are required, I will add that.

File: src/main/java/org/folio/services/UserService.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+   * @param username           username possibly used for login
+   * @param userId             user identifier possibly used for login
+   * @param tenant             effective tenant
+   * @param okapiURL           Okapi URL
+   * @param requestToken       request token
+   * @return asyncResult with user as a JsonObject
+   */
+  @Fluent
+  UserService lookupUser(String username, String userId, String tenant, final String okapiURL,
+      String requestToken, Handler<AsyncResult<JsonObject>> asyncResultHandler);
```
[the remark this answers is not in the document]
Comment (damien-git) [reply src/main/java/org/folio/services/UserService.java:45]: @Fluent requires returning the current class. @ProxyGen requires @Fluent. The vert.x [doc](https://vertx.io/docs/vertx-service-proxy/java/) gives examples without Fluent, but I don't know how to fix that.

[the remark this answers is not in the document]
Comment (damien-git) [reply src/main/java/org/folio/services/UserService.java:45]: Actually I think this is caused by the package-info.java. So I could remove it, but then I would have to fix all the other services.

[the remark this answers is not in the document]
Comment (damien-git) [reply src/main/java/org/folio/services/UserService.java:45]: OK, I will update all services.

File: ramls/userTenant.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply ramls/userTenant.json:25]: Tenant created by ops in okapi should be static, tenant name in consortia is to support customers, they can change as they want.
In some future, we want consortium admin to be able to add new tenant in consortia which will create tenant in okapi. At that time, the name of the tenant could be sync'ed between okapi and mod-consortia.
So as conclusion to this conversation we just need to delete tenantName from schema

File: src/main/java/org/folio/services/UserService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (damien-git) [reply src/main/java/org/folio/services/UserService.java:45]: Turns out this requires a lot of changes, and I don't have time to finish as I am going on vacation. I will create a separate PR when I come back.

File: src/main/java/org/folio/rest/impl/LoginAPI.java (hunk 5 above)
Comment (okolawole-ebsco) [reply src/main/java/org/folio/rest/impl/LoginAPI.java:264]: ok, i sorta get it now! could it be handleLogin?

File: src/main/java/org/folio/rest/impl/InitAPIs.java [hunk 10]
```diff
@@ -44,6 +46,9 @@ public void init(Vertx vertx, Context context, Handler<AsyncResult<Boolean>> res
...
+          .setAddress(MOD_USERS_PROXY_ADDRESS)
+          .register(UserService.class, UserService.create(vertx));
```
[the remark this answers is not in the document]
Comment (Dmitriy-Butramyou) [reply src/main/java/org/folio/rest/impl/InitAPIs.java]: fixed

File: src/main/java/org/folio/services/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+import org.folio.services.UserService;
+import org.folio.util.LoginConfigUtils;
+import org.folio.util.PercentCodec;
+import org.folio.util.StringUtil;
+import org.folio.util.WebClientFactory;
+
+import java.util.Map;
+
+public class UserServiceImpl implements UserService {
+  private static final String USER_TENANTS_URI_PATH = "/user-tenants";
```
[the remark this answers is not in the document]
Comment (Dmitriy-Butramyou) [reply src/main/java/org/folio/services/impl/UserServiceImpl.java:27]: Added

File: src/main/java/org/folio/services/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+      .map(response -> {
+        if (response.statusCode() != 200) {
+          String message = String.format(USER_TENANT_GET_ERROR, username, userId, requestedTenantId,
+            response.statusCode(), response.bodyAsString());
+          logger.info(message);
+          throw new RuntimeException(message);
+        }
+        JsonObject json = response.bodyAsJsonObject();
+        return json.getJsonArray("userTenants");
+      });
```
[the remark this answers is not in the document]
Comment (Dmitriy-Butramyou) [reply src/main/java/org/folio/services/impl/UserServiceImpl.java]: Refactored

File: src/main/java/org/folio/rest/impl/LoginAPI.java (hunk 5 above)
Comment (Dmitriy-Butramyou) [reply src/main/java/org/folio/rest/impl/LoginAPI.java]: Renamed

File: descriptors/ModuleDescriptor-template.json [hunk 13]
```diff
@@ -12,6 +12,7 @@
           "permissionsRequired" : [ "login.event.collection.post" ],
           "modulePermissions" : [
             "users.collection.get",
+            "user-tenants.collection.get",
```
[the remark this answers is not in the document]
Comment (Dmitriy-Butramyou) [reply descriptors/ModuleDescriptor-template.json]: Added

**Comment by NikitaSedyx** [comment]
hi @julianladisch and @steveellis looks like batch of comments are addressed, could you please review it again? thank you

File: src/main/java/org/folio/services/UserService.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+   * @param currentTenantId    identifier of the tenant being used in the login request
+   * @param username           username possibly used for login
+   * @param userId             user identifier possibly used for login
+   * @param requestedTenantId  identifier of the tenant requested by the user
+   * @param headers            okapi headers
+   * @return asyncResult with the response entity JsonArray of user-tenant associations
+   */
+  @Fluent
+  UserService getUserTenants(String currentTenantId, String username, String userId, String requestedTenantId,
+      JsonObject headers, Handler<AsyncResult<JsonArray>> asyncResultHandler);
```
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply src/main/java/org/folio/services/UserService.java:60]: It is written in the same style as an existing code, for refacroting all places to avoid Handler separate story will be created if needed

**Review by tatsiana-tarhonskaya — APPROVED** [review]

**Review by SinghAdes — APPROVED** [review]

**Review by Dmitriy-Butramyou — APPROVED** [review]
